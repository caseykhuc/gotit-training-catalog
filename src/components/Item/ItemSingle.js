import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Card } from 'react-bootstrap';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { getItem } from 'reducers';
import { fetchItem } from 'actions/item';
import { formatDateString } from 'utils';
import LoadingPage from 'components/common/LoadingPage';
import ModifyButton from 'components/common/ModifyButton';

export class ItemSingle extends Component {
  componentDidMount() {
    const {
      categoryId, itemId,
    } = this.props;
    this.fetchItem(categoryId, itemId);
  }

  componentDidUpdate(prevState) {
    const {
      categoryId, itemId,
    } = this.props;

    /* re-fetch item in case new item added
      when ItemSingle still mounted
    */
    if (categoryId !== prevState.categoryId
      || itemId !== prevState.itemId) {
      this.fetchItem(categoryId, itemId);
    }
  }

  fetchItem = async (categoryId, itemId) => {
    const { fetchItem } = this.props;
    const res = await fetchItem(categoryId, itemId);
    const { history } = this.props;
    if (!res.success) {
      history.push(`/categories/${categoryId}`)
    }
  }

  render() {
    const {
      item, userCurrent, categoryId, itemId, isLoadingItem,
    } = this.props;

    if (item) {
      const {
        id, name, description, price, userId, created, updated,
      } = item;

      return (
        <Card style={{
          width: '50%',
          margin: '0 auto',
        }}
        >
          <Card.Header>
            <h4>{`Item ${id}`}</h4>
            <small><i>{created && formatDateString(created)}</i></small>
          </Card.Header>
          <Card.Body className="text-center" style={{ paddingRight: '10%', paddingLeft: '10%' }}>

            <Card.Title>{name}</Card.Title>
            <Card.Text style={{ minHeight: '7rem', textAlign: 'left' }}>
              {description}
            </Card.Text>
            <Card.Text>
              <small className="text-muted">
                {updated && `Last updated: ${formatDateString(updated)}`}
              </small>
            </Card.Text>
            {userCurrent === userId && (
              <ModifyButton
                categoryId={categoryId}
                itemId={itemId}
                currentValue={{ name, description, price }}
              />
            )}
          </Card.Body>
        </Card>
      )
    }
    if (isLoadingItem) return <LoadingPage />;
    return '';
  }
}

export const mapStateToProps = (state, { match }) => {
  const { itemId, categoryId } = match.params;
  const { user, item } = state;
  return {
    itemId: Number(itemId),
    categoryId: Number(categoryId),
    item: getItem(state, itemId),
    userCurrent: user.userId,
    isLoadingItem: item.isLoading,
  }
}

ItemSingle.propTypes = {
  itemId: PropTypes.number.isRequired,
  categoryId: PropTypes.number.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  isLoadingItem: PropTypes.bool.isRequired,
  fetchItem: PropTypes.func.isRequired,
  item: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string,
    price: PropTypes.number.isRequired,
    userId: PropTypes.number.isRequired,
    created: PropTypes.string,
    updated: PropTypes.string,
  }),
  userCurrent: PropTypes.number,
}

export default withRouter(connect(mapStateToProps, { fetchItem })(ItemSingle));
