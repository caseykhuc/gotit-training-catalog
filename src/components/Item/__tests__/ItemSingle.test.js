import React from 'react';
import { shallow } from 'enzyme';
import { ButtonGroup } from 'react-bootstrap';
import { ItemSingle, mapStateToProps } from '../ItemSingle';

describe('component/Item/ItemSingle', () => {
  let props;
  let wrapper;
  let btnGroup;

  const update = () => {
    wrapper.update();
    btnGroup = wrapper.find(ButtonGroup);
  };
  const setup = () => {
    wrapper = shallow(<ItemSingle {...props} />);
    update();
  };

  beforeEach(() => {
    props = {
      itemId: 0,
      item: {
        name: 'test',
        description: 'test',
        id: 1,
        user_id: 0,
      },
      categoryId: 1,
      userCurrent: 10,
      fetchItem: jest.fn(),
    }
  });
  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('should render btnGroup when userId match', () => {
    props.item.user_id = 10;
    setup();
    expect(btnGroup.length).toBeTruthy();
  });
  it('should fetchItem when item is not defined in props', () => {
    props.item = undefined;
    setup();
    expect(props.fetchItem).toHaveBeenCalled();
  })
})

describe('component/Item/ItemSingle (mapStateToProps)', () => {
  let state; let match; let
    location;
  beforeEach(() => {
    state = { user: { userId: 10 }, item: { byId: [] } };
    match = { params: { categoryId: 1, itemId: 2 } };
  });

  it('should return the right categoryId/itemId from path', () => {
    expect(mapStateToProps(state, { match }).categoryId).toBe(1);
    expect(mapStateToProps(state, { match }).itemId).toBe(2);
  });
  it('should return userId/item correctly from state', () => {
    expect(mapStateToProps(state, { match }).userCurrent).toBe(10);
    expect(mapStateToProps(state, { match }).item).toBeFalsy();
  })
});
