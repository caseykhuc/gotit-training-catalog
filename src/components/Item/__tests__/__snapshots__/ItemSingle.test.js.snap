// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/Item/ItemSingle should render correctly 1`] = `ShallowWrapper {}`;

exports[`component/Item/ItemSingle should render correctly 2`] = `ShallowWrapper {}`;

exports[`component/Item/ItemSingle should render correctly 3`] = `ShallowWrapper {}`;

exports[`component/Item/ItemSingle should render correctly 4`] = `ShallowWrapper {}`;

exports[`component/Item/ItemSingle should render correctly 5`] = `ShallowWrapper {}`;
